### ISSUE TITLE
ICsvMapWriter sometimesfailed in multi threads

### ISSUE THREAD

**Issue by JunNakamura** [super-csv/super-csv#40]
I use super-csv 2.3.1 with play-2.2.1 on jdk-1.7.0-45.

I use ICsvWriter as the following to export some data as csv file.

```
mapWriter = new CsvMapWriter(osw, CsvPreference.EXCEL_PREFERENCE);
```

Firstly, I used older version (2.1.0). I faced the following bug, So I upgraded it to the current version.
http://sourceforge.net/p/supercsv/bugs/43/

It is OK at my local environment which is Windows7, memory is 16GB.
Then, I deployed my app to the staging environment, which is CentOS release 6.4 (Final), memory is 2GB.
 
But, StringIndexOutOfBoundsException sometimes occurred at the staging as below when running multiple threads.

```
Caused by: java.lang.StringIndexOutOfBoundsException: null
        at java.lang.AbstractStringBuilder.delete(AbstractStringBuilder.java:698) ~[na:1.7.0_45]
        at java.lang.StringBuilder.delete(StringBuilder.java:250) ~[na:1.7.0_45]
        at org.supercsv.encoder.DefaultCsvEncoder.encode(DefaultCsvEncoder.java:19) ~[myapp.myapp-1.0.0.0.jar:na]
        at org.supercsv.io.AbstractCsvWriter.writeRow(AbstractCsvWriter.java:177) ~[net.sf.supercsv.super-csv-2.3.1.jar:na]
        at org.supercsv.io.AbstractCsvWriter.writeHeader(AbstractCsvWriter.java:211) ~[net.sf.supercsv.super-csv-2.3.1.jar:na]
```

I try to create a DefaultCsvEncoder every time as the workaround of the bug 43, then the problem is solved.

```
CsvPreference preference =
                    new CsvPreference.Builder(CsvPreference.EXCEL_PREFERENCE)
                    .useEncoder(new DefaultCsvEncoder())
                    .build();
            mapWriter = new CsvMapWriter(osw, preference);
```
This workaround is acceptable for me, but I want to know the cause of this error if possible.
I wonder what is the difference my local and the staging.

**Comment by jamesbassett** [comment]
Sorry your post is a little ambiguous - it's not clear whether you still experience the bug in 2.3.1. This bug was fixed in [2.2.1](http://super-csv.github.io/super-csv/release_notes.html#a2.2.1), so any newer versions shouldn't experience this (if you are then please let us know!).

As for the reason? Each preference constant (e.g. `CsvPreference.EXCEL_PREFERENCE`) has an instance of `DefaultCsvEncoder`. Prior to this bug being fixed (see baef4f7ddcc9d528cdebe6847892d5842d813348), `DefaultCsvEncoder` had a `StringBuilder` instance field which was accessible via multiple threads.

**Comment by JunNakamura** [comment]
I must apologize... It is cause that my staging is illegal. 

I found that,  my colleague added org/supercsv/encoder/DefaultCsvEncoder.java, which version is 2.1.0,  to our package for debugging. The above is referred at the staging...

I removed it, then the problem is solved using 2.3.1. I will close this issue. I am deeply sorry　and thanks for helping.

**Comment by jamesbassett** [comment]
No worries Jun :)
